### ISSUE TITLE
Fix Nessus plugin update issue

### ISSUE THREAD

**Issue by pburkholder** [cloud-gov/cg-nessus-manager-boshrelease#36]
Since 10.x (or license change to 2048, and change back) broke the license registration on our node rebuilds.

Nessus support has not replied since Wednesday last week

**Comment by pburkholder** [comment]
This is open in Tenable at: https://community.tenable.com/s/case/5003a00001D0wpSAAR/reinstantiating-virtual-machine-breaks-plugin-updates

Case # 01346808

The gist of the matter, which references the script at: https://github.com/cloud-gov/cg-nessus-manager-boshrelease/blob/master/jobs/nessus-manager/templates/bin/ctl

---

To answer your question, the script seems to work successfully for every
step _except_ the attempt to run:

    $NESSUS_HOME/sbin/nessuscli fetch --register <%=p("nessus-manager.license") %>

And when I run it manually I get error message to the effect of:

'License already in use'

It seems to work for me only after I authenticate to tenable.com and then
reset the license in the webUI. Then I can manually run

nessuscli fetch --register ...

 This manual license reset step has only been necessary since we updated to
Nessus Manager 10.0. How can we restore the behavior we had for Nessus 8.X?

---

Reference also: https://twitter.com/pburkholder/status/1494502497876168706 which merited this comment:

**Comment by pburkholder** [comment]
That property is in credhub, at `credhub get -n /toolingbosh/nessus-manager-prod/nessus_manager_license`. So I can test in a shell on the node:

```
/opt/nessus/sbin/nessuscli fetch --register xxxx-xxxx-xxxx-xxxx-TP25
[error] Nessus Plugins: The server returned an error: License already in use
```

**Comment by pburkholder** [comment]
More notes:

According to Nessus_10_0.pdf, the command ` ./nessuscli fetch --register-only xxxx-xxxx-xxxx-xxxx-xxxx ` shouldn't return the "License already in use" error -- instead it should invalidate the previous license. I think that's how it worked in 8.x. Can we restore the prior behavior?

**Comment by pburkholder** [comment]
This matters because we can't get plugin updates:
```
# ./sbin/nessuscli update --plugins-only
Refreshing Nessus license information... complete; continuing with updates.


----- Fetching the newest updates from nessus.org -----

Could not validate this preference file.  Have installation files been copied from another system?
Nessus Plugins: Failed


 * Failed to update Nessus Plugins
```

**Comment by pburkholder** [comment]
This is still open as #01346808. Below is the latest email. I tried calling our old CSM but the number seems to route some service that does not seem like Tenable

--- 

to Support, cloud-gov-compliance@gsa.gov, szamora@tenable.com
Hi Jay,


You wrote:

     Also, we have not received any actionable data to determine the root cause. 


Please, do not say that unless you have bothered to review this case in full.

But let's go through this in detail. You wrote:
 

    If it's not the same host, you should not expect it to continue working.
    The activation code is only meant to work on one instance at a time

 
We've gone over this before, but I'm attaching the Nessus 10 PDF in full, as your support email wholly contradicts the product documentation, and included screenshots here:

First: this is the PDF for Nessus 10.1 from February 2022:

    image.png



Second, there's a section on "Transfer activation code" (highlights mine)

    image.png


Third, there are a set of command to transfer the code: (highlights mine)

    image.png



Fourth, I have run that command, unsuccessfully:

    image.png


But it gives me an unexpected error, and an invalid error as this instance is the only one running

Fifth: I stated this nature of this problem as clearly as possible on Feb 22 in an email to [support@tenable.com](mailto:support@tenable.com) on this same issue thread:

    image.png


and I referenced this exact same PDF on March 11:

    image.png




NOW -- What next?

1. It seems that we have encountered a bug in the Nessus product. It is not performing as documented
2. Is there a workaround for this bug, either at the server end (new license) or client patch?
3. Failing that, is there a timeline for fixing this bug?

Please include in your response Jeremy Newkirk, Sugey Zamora and our current CSM, as the support provided over the last two months has gotten us, from my perspective, nowhere.

I look forward to moving past this soon with a workable solution.

Peter Burkholder

**Comment by pburkholder** [comment]
I've also asked login-compliance and the gsait devsec ops folks if they have working contacts at Tenable.

**Comment by pburkholder** [comment]
The phone number for our Tenable CSM is now telling me on a recorded voice: "You might be able to save $200 on your car insurance"

**Comment by pburkholder** [comment]
The old CSM replied yesterday that "JR" will route this to the correct CSM. Still no substantive update.

**Comment by pburkholder** [comment]
> It's been another two days with no response. Can you please address my questions on how to run the bug report to generate the results you need?

**Comment by pburkholder** [comment]
Most recent timeline:

- Monday, April 11 - I sent a complete recap of the case
- Wednesday, April 13 - Jay of Tenable asks me for "bug report before it breaks and after it breaks" along with a link to their bug report documentation
- Wednesday, April 13 - I ask clarifying questions about what it means "before and after" 
- Friday, April 15 - I ask for an update on my clarifying questions
- Monday, April 18 - Tenable re-sends the April 13 documentation link, without answering any of my questions. *sigh*

**Comment by pburkholder** [comment]
* Monday, April 18 - I persisted past their malfunctioning website by using the Web Developer feature to replay some blocked requests, and I succeeded in uploading a redacted bug report. Their support staff didn't answer my question about the FISMA rating of their support system, so I had to redact hostnames with:
```perl -p -i.bak -e 's/[0-9a-f]{12}.accountid/redacted.redacted/' nessusd.*```

**Comment by pburkholder** [comment]
* Sunday, April 24 - We received a response after 6 days, which added no new information:
> According to the bug report the host is currently registered. 

*  Monday, April 25 - Sending new bug report for a host that is failing to register (which I probably should have done before...).  The process is
  * Download nessus-bug-report-20220425T154944Z.tar.gz, then: 
   ```
    tarfile=nessus-bug-report-20220425T154944Z.tar.gz
    mkdir nessus-bug
    cd nessus-bug
    tar zxf ../$tarfile
    cd bug-report
    perl -p -i.bak -e 's/[0-9a-f]{12}.accountid/redacted.redacted/' bug-report/nessusd.*
    rm  bug-report/*.bak
    tar zcvf ../$tarfile ./bug_report/
    ```

**Comment by pburkholder** [comment]
I made a mistake in my redaction process (literally using "accountid" in my Perl instead of actual account string) so now I've had to open P2-High for them to delete that file. 01386306 is the case for deletion of the non-redacted file. This is necessary since JAB does not want fully FQDN of internal hosts on non-authorized systems.

**Comment by pburkholder** [comment]
7 days without an update from Tenable. I asked again for an update
> Hi - It's been 7 days -- can you please update us with a path to resolving this? Thanks, Peter

**Comment by pburkholder** [comment]
Trying again since UI isn't working 

Can I update plugins? No:

```
$ sudo /opt/nessus/sbin/nessuscli update --plugins-only
Refreshing Nessus license information... complete; continuing with updates.


----- Fetching the newest updates from nessus.org -----

Could not validate this preference file.  Have installation files been copied from another system?
Nessus Plugins: Failed


 * Failed to update Nessus Plugins
 ```

Can I register? No, it thinks the license is in use:

```
$ sudo /opt/nessus/sbin/nessuscli fetch --register NHXX-XXXX-XXXX-XXXX-XXX5
[error] Nessus Plugins: The server returned an error: License already in use

```


So I reset the Product Activation in community.tenable.com:

![image](https://user-images.githubusercontent.com/1211146/219510284-ee9f8c2f-5204-4fea-b772-ffc81ead1387.png)


Then the fetch/register/update works:

```
$ sudo /opt/nessus/sbin/nessuscli fetch --register NXXX-XXXX-XXXX-XXXX-XXX5
Your Activation Code has been registered properly - thank you.
Refreshing Nessus license information... complete; continuing with updates.


----- Fetching the newest updates from nessus.org -----

Nessus Plugins: Downloading (66%)
Nessus Plugins: Unpacking (0%)
...
```

**Comment by pburkholder** [comment]
Hello Peter,

You shouldn't need to use the script every single time for plugin updates. The cause is happening because you are running the script instead of the intended normal method. This script is a hard reset update for when the normal method is not working as intended. (It's the reason you have to reset the activation code every single time) The auto-updater or manual update through GUI should be working just fine now and will update without needing to reset the activation code.

Image_2023-02-17_11-16-30.png

**Comment by pburkholder** [comment]
No reply in 3 days, so I wrote:
---
Hi,

It's been three days, how do you want to follow up on this?

Thanks, Peter

**Comment by pburkholder** [comment]
Nessus support followed up noting they didn't work on Monday, so I apologized for my eagerness and expressed my hope they had a good weekend and holiday.

They pointed out we need to have auto-plugin-updates set and so on. 

The next step here is to re-instantiate the Nessus manager to invoke the no-update issue. The command for that is:

> bosh -d nessus-manager-prod recreate nessus-manager

Here's my response, with only our screenshots included:

--
Good morning,

On Tue, Feb 21, 2023 at 3:21 PM Support <[support@tenable.com](mailto:support@tenable.com)> wrote:

> Hello Peter,
> I, unfortunately, don't work on the weekend and Monday was a holiday, sorry for the inconvenience. 

Thanks for the update, and I'm sorry to have implied you should be working 24x7 or such. I hope you had a good weekend and Presidents' Day

>From the looks of it, you are updating using the script. That script hard resets the license key and is not the recommended way to update your plugins.

I only went to the script because my attempts to update plugins in the UI wasn't working.

> Login to Nessus manager > settings > about > Software update and make sure automatic updates are enabled to update all components. Change the update frequency to whatever you want but we recommend daily.

Yes we have that set:

![image](https://user-images.githubusercontent.com/1211146/220702976-c30cfe32-2e29-4e20-9f5c-f4069609450a.png)

>Then after each day, you should see in the overview that it has been updated for the day. You should not need to run that script at all. If it runs into errors make sure the tenable websites are whitelisted and necessary ports are open
>   https://community.tenable.com/s/article/Which-Tenable-sites-should-I-whitelist
>   https://community.tenable.com/s/article/What-ports-are-required-for-Tenable-products
> Image_2023-02-21_13-17-53.png

Yes, this is _currently_ working because I already manually reset the license after our last O.S. update.

> Also, make sure to copy these advanced settings to make sure auto-update is enabled.
> Image_2023-02-21_13-21-02.png

Yes, we have these settings, except we manually update the Nessus version:

![image](https://user-images.githubusercontent.com/1211146/220703009-3b7730e6-2585-4ae2-9bca-6ae5f9790736.png)

![image](https://user-images.githubusercontent.com/1211146/220703039-6b5c3628-6431-4d3f-b5ac-3853cdd2fe63.png)

SO TO RECAP: We have the correct settings in place. In my next email I will detail what happens after we patch the O.S. and the system re-instantiates.

**Comment by pburkholder** [comment]
Nessus was quick to reply (yay!).

--

Looking forward to your next email. So when you update the OS of the scanner, it causes this issue? I misunderstood and assumed this issue happened during a core or Nessus update. I saw in your post that you were updating the scanner using the script which that script is a hard reset update that will reset the license. I didn't realize this error also happens when doing a simple OS update of the host running the scan. Curious to see what happens in detail 

--

My reply:

--

I simulated a patch by running a "recreate" process on our nessus-manager, which, like our patching process, :
a) destroys the existing (unpatched) VM for nessus-manager
b) instantiates a new fully-patched VM for nessus-manager
c) attaches the AWS EBS volume to the new VM

In this case, the plug updates seem to be running, and the Plugin Set is at 202302221414 (2022-02-22 14:14)

However, on the CLI the system seems confused about its license situation:

```
$ sudo /opt/nessus/sbin/nessuscli fetch --check
Checking...
Could not validate this preference file.  Have installation files been copied from another system?
This installation is not configured to receive updates

```
so we may have issues later. I will update if there are issues with the daily updates.

**Comment by pburkholder** [comment]
Hi Austin,	

> Hello Peter,
> That type of process will definitely need a new license as it practically rebuilds a new completely new instance. If this is how the update process of your systems is done, then it will, unfortunately, need to re-enter the license key every single time. 
> You could save a simple file to run automatically after the image is pushed like /opt/nessus/sbin/nessuscli fetch --register xxxx-xxxx-xxxx-xxxx  that you can simply run the command in conjunction with the new image. 

Yes, we have our instances use the register command after re-instantiating. You can see it right here in our bosh template:

> https://github.com/cloud-gov/cg-nessus-manager-boshrelease/blob/master/jobs/nessus-manager/templates/bin/ctl#L44
> `$NESSUS_HOME/sbin/nessuscli fetch --register <%= p("nessus-manager.license") %>`

that worked fine for 6 or 7 years. Now, however, we'll get something like the following:

```
/opt/nessus/sbin/nessuscli fetch --register xxxx-xxxx-xxxx-xxxx-TP25
[error] Nessus Plugins: The server returned an error: License already in use
```

So it seems that 10.X introduced a bug not present in Nessus 6, Nessus 7, or Nessus 8. How do we work around that?

Thanks,

**Comment by pburkholder** [comment]
Some really useful responses:

2023-02-28 / Thursday

--
 I spoke with the dev team and they explained why it worked in the past and why it doesn't work anymore. In the past when a license was in use it would automatically overwrite the license when the license activation command was run regardless if it was already in use or not. This caused many unfortunate problems where one employee would attempt to input the activation code on a different host and the original host would lose its license. The team believed that it would be better to pop up this warning message instead of instantly just running and overwriting the activation code. This warning was inputted so that if the license was in use it wouldn't automatically accept a license that was already activated and would need to manually reset the activation code as an additional confirmation step. Your way of updating the hosts is practically a new instance (not a simple yum update) so the new instance thinks the old instance is still the activation code so it pops up this error making sure you actually want to get rid of the activation code of the previous instance. 

I can see if they are interested in re-adding such a feature but I think they prefer the confirmation method so people don't accidentally use already activated activation codes causing the old host to lose the license. This can be a feature request and if lots of people agree with it, we may just reinstate the old method again if it's more desirable to a lot of people.

https://suggestions.tenable.com/

--

2023-02-28 / Monday

--

We do have an unregister command article but Nessus and community operate differently. Even if you reset the Nessus license manually, the community acts as a backup confirmation supposedly? It's supposed to prevent people from attempting to create multiple instances with the same product key and that's why an online community reset is required as well. I don't fully understand the reasoning for it as once you have unregistered your product manually, why would you also need to reset it in the community as well? I, unfortunately, don't make up these type of rules but maybe it will work. Try the manually reset registration and after your next update, see if the CLI register command works or if it's the same error code?

https://docs.tenable.com/nessus/commandlinereference/Content/ResetRegistrationAndEraseSettings.htm

The upper team told me it shouldn't work but I would like to test it at the very least as we don't have any api settings for you to use in community. They only told me that a feature request was needed to implement such a change but it doesn't seem like a difficult thing to add. Just make the previous unregister command also reset the activation code in community?

**Comment by pburkholder** [comment]
My latest reply to Tenable:

--

Hi Austin,

We're going to try the deregistration command you suggested.

If that doesn't work, I'd prefer to not go through the https://suggestions.tenable.com route for what is not a "feature request" - this is a bug that broke past functionality and I don't see that it should wait for votes from the other users. Is that something you can ask the product manager about?

Thanks, Peter

**Comment by pburkholder** [comment]
I wrote back:

--

What does it mean at https://docs.tenable.com/nessus/commandlinereference/Content/ResetRegistrationAndEraseSettings.htm when it says the following?

> Resetting Nessus configuration will permanently erase all your settings and cause Nessus to become unregistered.

I only want to unregister my system before recreating it. What are the "settings" as opposed to "all scans, scan data, policies, users and user settings, preferences and settings, registration information, and the master password" (from https://docs.tenable.com/nessus/commandlinereference/Content/ResetFull.htm)

**Comment by pburkholder** [comment]
We still have not identified a means for keeping Nessus plugins update following an update, but unassigning from me while on details.
